File: crates/cxx-qt-gen/test_outputs/types_primitive_property.cpp
#include "cxx-qt-gen/include/my_object.cxxqt.h"

namespace cxx_qt::my_object {

MyObject::MyObject(QObject* parent)
  : QObject(parent)
  , m_rustObj(cxx_qt::my_object::cxx_qt_my_object::createRs())
  , m_rustObjMutex(std::make_shared<std::mutex>())
  , m_cxxQtThreadObj(
      std::make_shared<rust::cxxqtlib1::CxxQtGuardedPointer<MyObject>>(this))
{
}

MyObject::~MyObject()
{
  const auto guard = std::unique_lock(m_cxxQtThreadObj->mutex);
  m_cxxQtThreadObj->ptr = nullptr;
}

const MyObjectRust&
MyObject::unsafeRust() const
{
  return *m_rustObj;
}

MyObjectRust&
MyObject::unsafeRustMut()
{
  return *m_rustObj;
}

std::unique_ptr<MyObjectCxxQtThread>
MyObject::qtThread() const
{
  return std::make_unique<MyObjectCxxQtThread>(m_cxxQtThreadObj,
                                               m_rustObjMutex);
}

const bool&
MyObject::getBoolean() const
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  return rust::cxxqtlib1::cxx_qt_convert<const bool&, const bool&>{}(
    m_rustObj->getBoolean(*this));
}

void
MyObject::emitBooleanChanged()
{
  const auto signalSuccess =
    QMetaObject::invokeMethod(this, "booleanChanged", Qt::QueuedConnection);
  Q_ASSERT(signalSuccess);
}

const float&
MyObject::getFloat32() const
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  return rust::cxxqtlib1::cxx_qt_convert<const float&, const float&>{}(
    m_rustObj->getFloat32(*this));
}

void
MyObject::emitFloat32Changed()
{
  const auto signalSuccess =
    QMetaObject::invokeMethod(this, "float32Changed", Qt::QueuedConnection);
  Q_ASSERT(signalSuccess);
}

const double&
MyObject::getFloat64() const
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  return rust::cxxqtlib1::cxx_qt_convert<const double&, const double&>{}(
    m_rustObj->getFloat64(*this));
}

void
MyObject::emitFloat64Changed()
{
  const auto signalSuccess =
    QMetaObject::invokeMethod(this, "float64Changed", Qt::QueuedConnection);
  Q_ASSERT(signalSuccess);
}

const qint8&
MyObject::getInt8() const
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  return rust::cxxqtlib1::cxx_qt_convert<const qint8&, const qint8&>{}(
    m_rustObj->getInt8(*this));
}

void
MyObject::emitInt8Changed()
{
  const auto signalSuccess =
    QMetaObject::invokeMethod(this, "int8Changed", Qt::QueuedConnection);
  Q_ASSERT(signalSuccess);
}

const qint16&
MyObject::getInt16() const
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  return rust::cxxqtlib1::cxx_qt_convert<const qint16&, const qint16&>{}(
    m_rustObj->getInt16(*this));
}

void
MyObject::emitInt16Changed()
{
  const auto signalSuccess =
    QMetaObject::invokeMethod(this, "int16Changed", Qt::QueuedConnection);
  Q_ASSERT(signalSuccess);
}

const qint32&
MyObject::getInt32() const
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  return rust::cxxqtlib1::cxx_qt_convert<const qint32&, const qint32&>{}(
    m_rustObj->getInt32(*this));
}

void
MyObject::emitInt32Changed()
{
  const auto signalSuccess =
    QMetaObject::invokeMethod(this, "int32Changed", Qt::QueuedConnection);
  Q_ASSERT(signalSuccess);
}

const quint8&
MyObject::getUint8() const
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  return rust::cxxqtlib1::cxx_qt_convert<const quint8&, const quint8&>{}(
    m_rustObj->getUint8(*this));
}

void
MyObject::emitUint8Changed()
{
  const auto signalSuccess =
    QMetaObject::invokeMethod(this, "uint8Changed", Qt::QueuedConnection);
  Q_ASSERT(signalSuccess);
}

const quint16&
MyObject::getUint16() const
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  return rust::cxxqtlib1::cxx_qt_convert<const quint16&, const quint16&>{}(
    m_rustObj->getUint16(*this));
}

void
MyObject::emitUint16Changed()
{
  const auto signalSuccess =
    QMetaObject::invokeMethod(this, "uint16Changed", Qt::QueuedConnection);
  Q_ASSERT(signalSuccess);
}

const quint32&
MyObject::getUint32() const
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  return rust::cxxqtlib1::cxx_qt_convert<const quint32&, const quint32&>{}(
    m_rustObj->getUint32(*this));
}

void
MyObject::emitUint32Changed()
{
  const auto signalSuccess =
    QMetaObject::invokeMethod(this, "uint32Changed", Qt::QueuedConnection);
  Q_ASSERT(signalSuccess);
}

void
MyObject::setBoolean(const bool& value)
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  m_rustObj->setBoolean(
    *this, rust::cxxqtlib1::cxx_qt_convert<bool, const bool&>{}(value));
}

void
MyObject::setFloat32(const float& value)
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  m_rustObj->setFloat32(
    *this, rust::cxxqtlib1::cxx_qt_convert<float, const float&>{}(value));
}

void
MyObject::setFloat64(const double& value)
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  m_rustObj->setFloat64(
    *this, rust::cxxqtlib1::cxx_qt_convert<double, const double&>{}(value));
}

void
MyObject::setInt8(const qint8& value)
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  m_rustObj->setInt8(
    *this, rust::cxxqtlib1::cxx_qt_convert<qint8, const qint8&>{}(value));
}

void
MyObject::setInt16(const qint16& value)
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  m_rustObj->setInt16(
    *this, rust::cxxqtlib1::cxx_qt_convert<qint16, const qint16&>{}(value));
}

void
MyObject::setInt32(const qint32& value)
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  m_rustObj->setInt32(
    *this, rust::cxxqtlib1::cxx_qt_convert<qint32, const qint32&>{}(value));
}

void
MyObject::setUint8(const quint8& value)
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  m_rustObj->setUint8(
    *this, rust::cxxqtlib1::cxx_qt_convert<quint8, const quint8&>{}(value));
}

void
MyObject::setUint16(const quint16& value)
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  m_rustObj->setUint16(
    *this, rust::cxxqtlib1::cxx_qt_convert<quint16, const quint16&>{}(value));
}

void
MyObject::setUint32(const quint32& value)
{
  const std::lock_guard<std::mutex> guard(*m_rustObjMutex);
  m_rustObj->setUint32(
    *this, rust::cxxqtlib1::cxx_qt_convert<quint32, const quint32&>{}(value));
}

} // namespace cxx_qt::my_object

namespace cxx_qt::my_object::cxx_qt_my_object {
std::unique_ptr<MyObject>
newCppObject()
{
  return std::make_unique<MyObject>();
}
} // namespace cxx_qt::my_object::cxx_qt_my_object

File: crates/cxx-qt-gen/test_outputs/types_primitive_property.rs
#[cxx::bridge(namespace = "cxx_qt::my_object")]
mod ffi {
    unsafe extern "C++" {
        include ! (< QtCore / QObject >);
        include!("cxx-qt-lib/include/convert.h");
        include!("cxx-qt-lib/include/cxxqt_thread.h");
    }

    unsafe extern "C++" {
        include!("cxx-qt-gen/include/my_object.cxxqt.h");
    }

    unsafe extern "C++" {
        #[cxx_name = "MyObject"]
        type MyObjectQt;

        #[rust_name = "emit_boolean_changed"]
        fn emitBooleanChanged(self: Pin<&mut MyObjectQt>);

        #[rust_name = "emit_float_32_changed"]
        fn emitFloat32Changed(self: Pin<&mut MyObjectQt>);

        #[rust_name = "emit_float_64_changed"]
        fn emitFloat64Changed(self: Pin<&mut MyObjectQt>);

        #[rust_name = "emit_int_8_changed"]
        fn emitInt8Changed(self: Pin<&mut MyObjectQt>);

        #[rust_name = "emit_int_16_changed"]
        fn emitInt16Changed(self: Pin<&mut MyObjectQt>);

        #[rust_name = "emit_int_32_changed"]
        fn emitInt32Changed(self: Pin<&mut MyObjectQt>);

        #[rust_name = "emit_uint_8_changed"]
        fn emitUint8Changed(self: Pin<&mut MyObjectQt>);

        #[rust_name = "emit_uint_16_changed"]
        fn emitUint16Changed(self: Pin<&mut MyObjectQt>);

        #[rust_name = "emit_uint_32_changed"]
        fn emitUint32Changed(self: Pin<&mut MyObjectQt>);
    }

    extern "Rust" {
        #[cxx_name = "MyObjectRust"]
        type MyObject;

        #[cxx_name = "getBoolean"]
        unsafe fn get_boolean<'a>(self: &'a MyObject, cpp: &'a MyObjectQt) -> &'a bool;
        #[cxx_name = "setBoolean"]
        fn set_boolean(self: &mut MyObject, cpp: Pin<&mut MyObjectQt>, value: bool);

        #[cxx_name = "getFloat32"]
        unsafe fn get_float_32<'a>(self: &'a MyObject, cpp: &'a MyObjectQt) -> &'a f32;
        #[cxx_name = "setFloat32"]
        fn set_float_32(self: &mut MyObject, cpp: Pin<&mut MyObjectQt>, value: f32);

        #[cxx_name = "getFloat64"]
        unsafe fn get_float_64<'a>(self: &'a MyObject, cpp: &'a MyObjectQt) -> &'a f64;
        #[cxx_name = "setFloat64"]
        fn set_float_64(self: &mut MyObject, cpp: Pin<&mut MyObjectQt>, value: f64);

        #[cxx_name = "getInt8"]
        unsafe fn get_int_8<'a>(self: &'a MyObject, cpp: &'a MyObjectQt) -> &'a i8;
        #[cxx_name = "setInt8"]
        fn set_int_8(self: &mut MyObject, cpp: Pin<&mut MyObjectQt>, value: i8);

        #[cxx_name = "getInt16"]
        unsafe fn get_int_16<'a>(self: &'a MyObject, cpp: &'a MyObjectQt) -> &'a i16;
        #[cxx_name = "setInt16"]
        fn set_int_16(self: &mut MyObject, cpp: Pin<&mut MyObjectQt>, value: i16);

        #[cxx_name = "getInt32"]
        unsafe fn get_int_32<'a>(self: &'a MyObject, cpp: &'a MyObjectQt) -> &'a i32;
        #[cxx_name = "setInt32"]
        fn set_int_32(self: &mut MyObject, cpp: Pin<&mut MyObjectQt>, value: i32);

        #[cxx_name = "getUint8"]
        unsafe fn get_uint_8<'a>(self: &'a MyObject, cpp: &'a MyObjectQt) -> &'a u8;
        #[cxx_name = "setUint8"]
        fn set_uint_8(self: &mut MyObject, cpp: Pin<&mut MyObjectQt>, value: u8);

        #[cxx_name = "getUint16"]
        unsafe fn get_uint_16<'a>(self: &'a MyObject, cpp: &'a MyObjectQt) -> &'a u16;
        #[cxx_name = "setUint16"]
        fn set_uint_16(self: &mut MyObject, cpp: Pin<&mut MyObjectQt>, value: u16);

        #[cxx_name = "getUint32"]
        unsafe fn get_uint_32<'a>(self: &'a MyObject, cpp: &'a MyObjectQt) -> &'a u32;
        #[cxx_name = "setUint32"]
        fn set_uint_32(self: &mut MyObject, cpp: Pin<&mut MyObjectQt>, value: u32);
    }

    unsafe extern "C++" {
        type MyObjectCxxQtThread;

        #[cxx_name = "unsafeRust"]
        fn rust(self: &MyObjectQt) -> &MyObject;

        #[cxx_name = "qtThread"]
        fn qt_thread(self: &MyObjectQt) -> UniquePtr<MyObjectCxxQtThread>;
        fn queue(self: &MyObjectCxxQtThread, func: fn(ctx: Pin<&mut MyObjectQt>)) -> Result<()>;

        #[rust_name = "new_cpp_object_my_object_qt"]
        #[namespace = "cxx_qt::my_object::cxx_qt_my_object"]
        fn newCppObject() -> UniquePtr<MyObjectQt>;
    }

    extern "C++" {
        #[cxx_name = "unsafeRustMut"]
        unsafe fn rust_mut(self: Pin<&mut MyObjectQt>) -> Pin<&mut MyObject>;
    }

    extern "Rust" {
        #[cxx_name = "createRs"]
        #[namespace = "cxx_qt::my_object::cxx_qt_my_object"]
        fn create_rs_my_object() -> Box<MyObject>;
    }
}

pub use self::cxx_qt_ffi::*;
mod cxx_qt_ffi {
    use super::ffi::*;
    use std::pin::Pin;

    type UniquePtr<T> = cxx::UniquePtr<T>;

    #[derive(Default)]
    pub struct MyObject {
        boolean: bool,
        float_32: f32,
        float_64: f64,
        int_8: i8,
        int_16: i16,
        int_32: i32,
        uint_8: u8,
        uint_16: u16,
        uint_32: u32,
    }

    impl MyObject {
        pub fn get_boolean<'a>(&'a self, cpp: &'a MyObjectQt) -> &'a bool {
            cpp.get_boolean()
        }

        pub fn set_boolean(&mut self, cpp: Pin<&mut MyObjectQt>, value: bool) {
            cpp.set_boolean(value);
        }

        pub fn get_float_32<'a>(&'a self, cpp: &'a MyObjectQt) -> &'a f32 {
            cpp.get_float_32()
        }

        pub fn set_float_32(&mut self, cpp: Pin<&mut MyObjectQt>, value: f32) {
            cpp.set_float_32(value);
        }

        pub fn get_float_64<'a>(&'a self, cpp: &'a MyObjectQt) -> &'a f64 {
            cpp.get_float_64()
        }

        pub fn set_float_64(&mut self, cpp: Pin<&mut MyObjectQt>, value: f64) {
            cpp.set_float_64(value);
        }

        pub fn get_int_8<'a>(&'a self, cpp: &'a MyObjectQt) -> &'a i8 {
            cpp.get_int_8()
        }

        pub fn set_int_8(&mut self, cpp: Pin<&mut MyObjectQt>, value: i8) {
            cpp.set_int_8(value);
        }

        pub fn get_int_16<'a>(&'a self, cpp: &'a MyObjectQt) -> &'a i16 {
            cpp.get_int_16()
        }

        pub fn set_int_16(&mut self, cpp: Pin<&mut MyObjectQt>, value: i16) {
            cpp.set_int_16(value);
        }

        pub fn get_int_32<'a>(&'a self, cpp: &'a MyObjectQt) -> &'a i32 {
            cpp.get_int_32()
        }

        pub fn set_int_32(&mut self, cpp: Pin<&mut MyObjectQt>, value: i32) {
            cpp.set_int_32(value);
        }

        pub fn get_uint_8<'a>(&'a self, cpp: &'a MyObjectQt) -> &'a u8 {
            cpp.get_uint_8()
        }

        pub fn set_uint_8(&mut self, cpp: Pin<&mut MyObjectQt>, value: u8) {
            cpp.set_uint_8(value);
        }

        pub fn get_uint_16<'a>(&'a self, cpp: &'a MyObjectQt) -> &'a u16 {
            cpp.get_uint_16()
        }

        pub fn set_uint_16(&mut self, cpp: Pin<&mut MyObjectQt>, value: u16) {
            cpp.set_uint_16(value);
        }

        pub fn get_uint_32<'a>(&'a self, cpp: &'a MyObjectQt) -> &'a u32 {
            cpp.get_uint_32()
        }

        pub fn set_uint_32(&mut self, cpp: Pin<&mut MyObjectQt>, value: u32) {
            cpp.set_uint_32(value);
        }
    }

    impl MyObjectQt {
        pub fn get_boolean(&self) -> &bool {
            &self.rust().boolean
        }

        pub fn set_boolean(mut self: Pin<&mut Self>, value: bool) {
            unsafe {
                self.as_mut().rust_mut().boolean = value;
            }
            self.as_mut().emit_boolean_changed();
        }

        pub fn get_float_32(&self) -> &f32 {
            &self.rust().float_32
        }

        pub fn set_float_32(mut self: Pin<&mut Self>, value: f32) {
            unsafe {
                self.as_mut().rust_mut().float_32 = value;
            }
            self.as_mut().emit_float_32_changed();
        }

        pub fn get_float_64(&self) -> &f64 {
            &self.rust().float_64
        }

        pub fn set_float_64(mut self: Pin<&mut Self>, value: f64) {
            unsafe {
                self.as_mut().rust_mut().float_64 = value;
            }
            self.as_mut().emit_float_64_changed();
        }

        pub fn get_int_8(&self) -> &i8 {
            &self.rust().int_8
        }

        pub fn set_int_8(mut self: Pin<&mut Self>, value: i8) {
            unsafe {
                self.as_mut().rust_mut().int_8 = value;
            }
            self.as_mut().emit_int_8_changed();
        }

        pub fn get_int_16(&self) -> &i16 {
            &self.rust().int_16
        }

        pub fn set_int_16(mut self: Pin<&mut Self>, value: i16) {
            unsafe {
                self.as_mut().rust_mut().int_16 = value;
            }
            self.as_mut().emit_int_16_changed();
        }

        pub fn get_int_32(&self) -> &i32 {
            &self.rust().int_32
        }

        pub fn set_int_32(mut self: Pin<&mut Self>, value: i32) {
            unsafe {
                self.as_mut().rust_mut().int_32 = value;
            }
            self.as_mut().emit_int_32_changed();
        }

        pub fn get_uint_8(&self) -> &u8 {
            &self.rust().uint_8
        }

        pub fn set_uint_8(mut self: Pin<&mut Self>, value: u8) {
            unsafe {
                self.as_mut().rust_mut().uint_8 = value;
            }
            self.as_mut().emit_uint_8_changed();
        }

        pub fn get_uint_16(&self) -> &u16 {
            &self.rust().uint_16
        }

        pub fn set_uint_16(mut self: Pin<&mut Self>, value: u16) {
            unsafe {
                self.as_mut().rust_mut().uint_16 = value;
            }
            self.as_mut().emit_uint_16_changed();
        }

        pub fn get_uint_32(&self) -> &u32 {
            &self.rust().uint_32
        }

        pub fn set_uint_32(mut self: Pin<&mut Self>, value: u32) {
            unsafe {
                self.as_mut().rust_mut().uint_32 = value;
            }
            self.as_mut().emit_uint_32_changed();
        }
    }

    unsafe impl Send for MyObjectCxxQtThread {}

    pub fn create_rs_my_object() -> std::boxed::Box<MyObject> {
        std::default::Default::default()
    }
}
